readme.md
Joystick Plugin for CryEngine
=============================
Adds Flownode for Joystick and Controller Input.
Port of the FGPS Joystick Node to the Plugin SDK.

Installation / Integration
==========================
Use the installer or extract the files to your Cryengine SDK Folder so that the Code and BinXX/Plugins directory match up.

The plugin manager will automatically load up the plugin when the game/editor is restarted or if you directly load it.

Flownodes
=========
```Plugin_Joystick:Input``` Joystick Input
* In ```Enabled``` Enables/Disables the joystick
* In ```Joystick``` Select the joystick
* Out ```X``` X-axis, usually the left-right movement of a stick.
* Out ```Y``` Y-axis, usually the forward-backward movement of a stick.
* Out ```Z``` Z-axis, often the throttle control. If the joystick does not have this axis, the value is 0.
* Out ```X-Rotate``` X-axis rotation. If the joystick does not have this axis, the value is 0.
* Out ```Y-Rotate``` Y-axis rotation. If the joystick does not have this axis, the value is 0.
* Out ```Z-Rotate``` Z-axis rotation (often called the rudder). If the joystick does not have this axis, the value is 0.
* Out ```Slider``` Two additional axis values (formerly called the u-axis and v-axis) whose semantics depend on the joystick.
* Out ```POV``` Direction controllers, such as point-of-view hats.
  The position is indicated in hundredths of a degree clockwise from north (away from the user).
  The center position is normally reported as -1.
  For indicators that have only five positions, the value for a controller is -1, 0, 9,000, 18,000, or 27,000.
* Out ```ButtonPressed``` Joystick button pressed
* Out ```ButtonReleased``` Joystick button released

changelog.md
Joystick Plugin 1.2.2.0 (29.11.2013)
---------------------------
Stable Release for CryEngine 3.5.4 (32/64 bit)

Joystick Plugin 1.2.1.0 (02.03.2013)
---------------------------
Stable Release for CryEngine 3.4.5 (32/64 bit)

Joystick Plugin 1.2.0.0 (08.02.2012)
---------------------------
Stable Release for CryEngine 3.4.4 (32/64 bit)

Joystick Plugin 1.1.0.0 (19.11.2012)
---------------------------
Stable Release for CryEngine 3.4.3 (32/64 bit)

Joystick Plugin 1.0.0.0 (30.10.2012)
---------------------------
Initial Stable Release for CryEngine 3.4.0 (32/64 bit)


src/StdAfx.h
/* Joystick_Plugin - for licensing and copyright see license.txt */

#pragma once

// Insert your headers here
#include <platform.h>
#include <algorithm>
#include <vector>
#include <memory>
#include <list>
#include <functional>
#include <limits>
#include <smartptr.h>
#include <CryThread.h>
#include <Cry_Math.h>
#include <ISystem.h>
#include <I3DEngine.h>
#include <IInput.h>
#include <IConsole.h>
#include <ITimer.h>
#include <ILog.h>
#include <IGameplayRecorder.h>
#include <ISerialize.h>

#ifndef _FORCEDLL
#define _FORCEDLL
#endif

#ifndef JOYSTICKPLUGIN_EXPORTS
#define JOYSTICKPLUGIN_EXPORTS
#endif

#pragma warning(disable: 4018)  // conditional expression is constant

//{{AFX_INSERT_LOCATION}}
// Microsoft Visual C++ will insert additional declarations immediately before the previous line.


src/CPluginJoystick.h
/* Joystick_Plugin - for licensing and copyright see license.txt */

#pragma once

#include <IPluginBase.h>
#include <Game.h>

#include <IPluginManager.h>
#include <CPluginBase.hpp>

#include <IPluginJoystick.h>

#include <dinput.h>

#define PLUGIN_NAME "Joystick"
#define PLUGIN_CONSOLE_PREFIX "[" PLUGIN_NAME " " PLUGIN_TEXT "] " //!< Prefix for Logentries by this plugin

namespace JoystickPlugin
{
    extern void* g_hInst;
    extern LPDIRECTINPUT8 g_lpdi;
    typedef std::map<LPDIRECTINPUTDEVICE8, DIDEVCAPS> TJoysticks;
    extern TJoysticks g_joysticks;

    /**
    * @brief Provides information and manages the resources of this plugin.
    */
    class CPluginJoystick :
        public PluginManager::CPluginBase,
        public IPluginJoystick
    {
        public:
            CPluginJoystick();
            ~CPluginJoystick();

            // IPluginBase
            bool Release( bool bForce = false );

            int GetInitializationMode() const
            {
                return int( PluginManager::IM_Default );
            };

            bool Init( SSystemGlobalEnvironment& env, SSystemInitParams& startupParams, IPluginBase* pPluginManager, const char* sPluginDirectory );

            const char* GetVersion() const
            {
                return "1.2.2.0";
            };

            const char* GetName() const
            {
                return PLUGIN_NAME;
            };

            const char* GetCategory() const
            {
                return "Hardware";
            };

            const char* ListAuthors() const
            {
                return "James-Ryan <neo-ryan at web dot de>"
                       ",\nHendrik Polczynski - Plugin SDK Port <hendrikpolczyn at gmail dot com>";
            };

            const char* ListCVars() const;

            const char* GetStatus() const;

            const char* GetCurrentConcreteInterfaceVersion() const
            {
                return "1.0";
            };

            void* GetConcreteInterface( const char* sInterfaceVersion )
            {
                return static_cast <IPluginJoystick*>( this );
            };

            // IPluginJoystick
            IPluginBase* GetBase()
            {
                return static_cast<IPluginBase*>( this );
            };

            // TODO: Add your concrete interface implementation
    };

    extern CPluginJoystick* gPlugin;
}

/**
* @brief This function is required to use the Autoregister Flownode without modification.
* Include the file "CPluginJoystick.h" in front of flownode.
*/
inline void GameWarning( const char* sFormat, ... ) PRINTF_PARAMS( 1, 2 );
inline void GameWarning( const char* sFormat, ... )
{
    va_list ArgList;
    va_start( ArgList, sFormat );
    JoystickPlugin::gPlugin->LogV( ILog::eWarningAlways, sFormat, ArgList );
    va_end( ArgList );
};


src/CPluginJoystick.cpp
/* Joystick_Plugin - for licensing and copyright see license.txt */

#include <StdAfx.h>
#include <CPluginJoystick.h>

#include <dinput.h>

namespace JoystickPlugin
{
    CPluginJoystick* gPlugin = NULL;

    void* g_hInst = 0;
    LPDIRECTINPUT8 g_lpdi = NULL;
    TJoysticks g_joysticks;
    LPDIRECTINPUTDEVICE8 currentJoystick;

    BOOL CALLBACK EnumAxesCallback( const DIDEVICEOBJECTINSTANCE* instance, VOID* context )
    {
        HWND hDlg = ( HWND )context;

        DIPROPRANGE propRange;
        propRange.diph.dwSize       = sizeof( DIPROPRANGE );
        propRange.diph.dwHeaderSize = sizeof( DIPROPHEADER );
        propRange.diph.dwHow        = DIPH_BYID;
        propRange.diph.dwObj        = instance->dwType;
        propRange.lMin              = -1000;
        propRange.lMax              = +1000;

        if ( FAILED( currentJoystick->SetProperty( DIPROP_RANGE, &propRange.diph ) ) )
        {
            return DIENUM_STOP;
        }

        return DIENUM_CONTINUE;
    }

    BOOL CALLBACK EnumDevicesCallback( const DIDEVICEINSTANCE* instance, VOID* context )
    {
        HRESULT hr;

        LPDIRECTINPUTDEVICE8 joystick;
        hr = g_lpdi->CreateDevice( instance->guidInstance, &joystick, NULL );
        currentJoystick = joystick;

        if ( FAILED( hr ) )
        {
            return DIENUM_CONTINUE;
        }

        if ( FAILED( hr = joystick->SetDataFormat( &c_dfDIJoystick2 ) ) )
        {
            return DIENUM_CONTINUE;
        }

        DIDEVCAPS capabilities;
        capabilities.dwSize = sizeof( DIDEVCAPS );

        if ( FAILED( hr = joystick->GetCapabilities( &capabilities ) ) )
        {
            return DIENUM_CONTINUE;
        }

        if ( FAILED( hr = joystick->EnumObjects( EnumAxesCallback, NULL, DIDFT_AXIS ) ) )
        {
            return DIENUM_CONTINUE;
        }

        g_joysticks.insert( TJoysticks::value_type( joystick, capabilities ) );

        if ( FAILED( hr = joystick->SetCooperativeLevel( NULL, DISCL_EXCLUSIVE | DISCL_FOREGROUND ) ) )
        {
            return DIENUM_CONTINUE;
        }

        return DIENUM_STOP;
    }

    HRESULT PollJoystick( DIJOYSTATE2* js, LPDIRECTINPUTDEVICE8 joystick )
    {
        HRESULT hr;

        if ( joystick == NULL )
        {
            return S_OK;
        }

        hr = joystick->Poll();

        if ( FAILED( hr ) )
        {

            hr = joystick->Acquire();

            while ( hr == DIERR_INPUTLOST )
            {
                hr = joystick->Acquire();
            }

            if ( ( hr == DIERR_INVALIDPARAM ) || ( hr == DIERR_NOTINITIALIZED ) )
            {
                return E_FAIL;
            }

            if ( hr == DIERR_OTHERAPPHASPRIO )
            {
                return S_OK;
            }
        }

        if ( FAILED( hr = joystick->GetDeviceState( sizeof( DIJOYSTATE2 ), js ) ) )
        {
            return hr;
        }

        return S_OK;
    }


    CPluginJoystick::CPluginJoystick()
    {
        gPlugin = this;
    }

    CPluginJoystick::~CPluginJoystick()
    {
        Release( true );

        gPlugin = NULL;
    }

    bool CPluginJoystick::Release( bool bForce )
    {
        bool bRet = true;

        if ( !m_bCanUnload )
        {
            // Should be called while Game is still active otherwise there might be leaks/problems
            bRet = CPluginBase::Release( bForce );

            if ( bRet )
            {
                // Depending on your plugin you might not want to unregister anything
                // if the System is quitting.
                if ( gEnv && gEnv->pSystem && !gEnv->pSystem->IsQuitting() )
                {
                    // Unregister CVars
                    if ( gEnv && gEnv->pConsole )
                    {
                        // ...
                    }

                    // Unregister game objects
                    if ( gEnv && gEnv->pGame && gEnv->pGame->GetIGameFramework() )
                    {
                        // ...
                    }
                }

                // Cleanup like this always (since the class is static its cleaned up when the dll is unloaded)
                gPluginManager->UnloadPlugin( GetName() );

                // Allow Plugin Manager garbage collector to unload this plugin
                AllowDllUnload();
            }
        }

        return bRet;
    };

    bool CPluginJoystick::Init( SSystemGlobalEnvironment& env, SSystemInitParams& startupParams, IPluginBase* pPluginManager, const char* sPluginDirectory )
    {
        gPluginManager = ( PluginManager::IPluginManager* )pPluginManager->GetConcreteInterface( NULL );
        CPluginBase::Init( env, startupParams, pPluginManager, sPluginDirectory );

        if ( gEnv && gEnv->pSystem && !gEnv->pSystem->IsQuitting() )
        {
            if ( DirectInput8Create( static_cast<HINSTANCE>( g_hInst ), DIRECTINPUT_VERSION, IID_IDirectInput8, reinterpret_cast<void**>( &g_lpdi ), NULL ) == DI_OK )
            {
                g_lpdi->EnumDevices( DI8DEVCLASS_GAMECTRL, EnumDevicesCallback, NULL, DIEDFL_ATTACHEDONLY );
            }

            else
            {
                gPlugin->LogError( "[DIRECTINPUT] Failed to create input device" );
            }
        }

        // Note: Autoregister Flownodes will be automatically registered

        return true;
    }

    const char* CPluginJoystick::ListCVars() const
    {
        return "..."; // TODO: Enter CVARs/Commands here if you have some
    }

    const char* CPluginJoystick::GetStatus() const
    {
        return "OK";
    }

    // TODO: Add your plugin concrete interface implementation
}

src/CPluginJoystickModule.cpp
/* Joystick_Plugin - for licensing and copyright see license.txt */

#include <StdAfx.h>
#include <IPluginManager.h>
#include "CPluginJoystick.h"
#include <platform_impl.h>
#include <Nodes/G2FlowBaseNode.h>

extern "C"
{
    DLL_EXPORT PluginManager::IPluginBase* GetPluginInterface( const char* sInterfaceVersion )
    {
        // This function should not create a new interface class each call.
        static JoystickPlugin::CPluginJoystick modulePlugin;
        JoystickPlugin::gPlugin = &modulePlugin;
        return modulePlugin.GetBase();
    }
}

PluginManager::IPluginManager* gPluginManager = NULL; //!< pointer to plugin manager

// Needed for module specific flow node
CG2AutoRegFlowNodeBase* CG2AutoRegFlowNodeBase::m_pFirst = 0; //!< pointer to first flownode inside this plugin
CG2AutoRegFlowNodeBase* CG2AutoRegFlowNodeBase::m_pLast = 0; //!< pointer to last flownode inside this plugin

#define WIN32_LEAN_AND_MEAN
#include <Windows.h>

/**
* @brief DLL Entry point.
* Has no function(yet?) in this plugin system.
*/
BOOL APIENTRY DllMain( HMODULE hModule, DWORD ul_reason_for_call, LPVOID lpReserved )
{
    switch ( ul_reason_for_call )
    {
        case DLL_PROCESS_ATTACH:
            JoystickPlugin::g_hInst = hModule;
            break;

        case DLL_THREAD_ATTACH:
            break;

        case DLL_THREAD_DETACH:
            break;

        case DLL_PROCESS_DETACH:
            break;
    }

    return TRUE;
}

src/Flownodes/FlowJoystick.cpp
/* Joystick_Plugin - for licensing and copyright see license.txt */

#include "StdAfx.h"

#include <CPluginJoystick.h>
#include "Nodes/G2FlowBaseNode.h"

namespace JoystickPlugin
{
    HRESULT PollJoystick( DIJOYSTATE2* js, LPDIRECTINPUTDEVICE8 joystick );

    class CFlowJoystickNode :
        public CFlowBaseNode<eNCT_Instanced>,
        public CScriptableBase
    {
        private:

            SActivationInfo m_actInfo;
            int* m_buttons;
            int m_xAxis;
            int m_yAxis;
            int m_zAxis;
            int m_xRotate;
            int m_yRotate;
            int m_zRotate;
            int m_slider;
            int m_POV;
            bool m_bEnabled;

            LPDIRECTINPUTDEVICE8 m_currentJoystick;
            DIDEVCAPS m_CurrentCapabilities;

        public:

            CFlowJoystickNode( SActivationInfo* pActInfo )
            {
#if CDK_VERSION < 343 || CDK_VERSION >= 350
                Init( gEnv->pSystem->GetIScriptSystem(), gEnv->pSystem );
#else
                Init( gEnv->pSystem->GetIScriptSystem() );
#endif
                SetGlobalName( "Joystick" );

#undef SCRIPT_REG_CLASSNAME
#define SCRIPT_REG_CLASSNAME &CFlowJoystickNode::

                SCRIPT_REG_TEMPLFUNC( GetXAxis, "" );
                SCRIPT_REG_TEMPLFUNC( GetYAxis, "" );
                SCRIPT_REG_TEMPLFUNC( GetZAxis, "" );
                SCRIPT_REG_TEMPLFUNC( GetButtonState, "button" );
                SCRIPT_REG_TEMPLFUNC( SetJoystick, "joystick" );
                SCRIPT_REG_TEMPLFUNC( GetNumButtons, "" );
                SCRIPT_REG_TEMPLFUNC( GetPOV, "" );
                SCRIPT_REG_TEMPLFUNC( GetSlider, "" );
                SCRIPT_REG_TEMPLFUNC( GetXRotate, "" );
                SCRIPT_REG_TEMPLFUNC( GetYRotate, "" );
                SCRIPT_REG_TEMPLFUNC( GetZRotate, "" );

#undef SCRIPT_REG_CLASSNAME

                m_xAxis = 0;
                m_yAxis = 0;
                m_zAxis = 0;
                m_xRotate = 0;
                m_yRotate = 0;
                m_zRotate = 0;
                m_slider = 0;
                m_POV = 0;
                m_buttons = NULL;
                m_currentJoystick = NULL;
                m_bEnabled = true;
            }

            ~CFlowJoystickNode()
            {
                SAFE_DELETE( m_buttons );

                if ( m_currentJoystick )
                {
                    m_currentJoystick->Unacquire();
                }
            }

            int CFlowJoystickNode::GetButtonState( IFunctionHandler* pH, int button )
            {
                if ( !m_buttons || button > m_CurrentCapabilities.dwButtons || button < 1 )
                {
                    return pH->EndFunction();
                }

                return pH->EndFunction( m_buttons[button - 1] );
            }

            int CFlowJoystickNode::GetNumButtons( IFunctionHandler* pH )
            {
                return pH->EndFunction( ( int )m_CurrentCapabilities.dwButtons );
            }

            int CFlowJoystickNode::SetJoystick( IFunctionHandler* pH, int joystick )
            {
                if ( joystick > g_joysticks.size() || joystick < 1 )
                {
                    SetNewJoystick( 1 );
                    return pH->EndFunction();
                }

                SetNewJoystick( joystick );

                return pH->EndFunction();
            }

            int CFlowJoystickNode::GetXAxis( IFunctionHandler* pH )
            {
                return pH->EndFunction( m_xAxis );
            }

            int CFlowJoystickNode::GetYAxis( IFunctionHandler* pH )
            {
                return pH->EndFunction( m_yAxis );
            }

            int CFlowJoystickNode::GetZAxis( IFunctionHandler* pH )
            {
                return pH->EndFunction( m_zAxis );
            }

            int CFlowJoystickNode::GetXRotate( IFunctionHandler* pH )
            {
                return pH->EndFunction( m_xRotate );
            }

            int CFlowJoystickNode::GetYRotate( IFunctionHandler* pH )
            {
                return pH->EndFunction( m_yRotate );
            }

            int CFlowJoystickNode::GetZRotate( IFunctionHandler* pH )
            {
                return pH->EndFunction( m_zRotate );
            }

            int CFlowJoystickNode::GetPOV( IFunctionHandler* pH )
            {
                return pH->EndFunction( m_POV );
            }

            int CFlowJoystickNode::GetSlider( IFunctionHandler* pH )
            {
                return pH->EndFunction( m_slider );
            }

            IFlowNodePtr Clone( SActivationInfo* pActInfo )
            {
                return new CFlowJoystickNode( pActInfo );
            }

            virtual void GetMemoryUsage( ICrySizer* s ) const
            {
                s->Add( *this );
            }

            void Serialize( SActivationInfo* pActInfo, TSerialize ser )
            {
                if ( ser.IsReading() )
                {
                    m_actInfo = *pActInfo;
                }
            }

            enum EInputPorts
            {
                EIP_Enabled = 0,
                EIP_JoystickSelection
            };

            enum EOutputPorts
            {
                EOP_JoystickX = 0,
                EOP_JoystickY,
                EOP_JoystickZ,
                EOP_JoystickRotateX,
                EOP_JoystickRotateY,
                EOP_JoystickRotateZ,
                EOP_Slider,
                EOP_POV,
                EOP_ButtonPressed,
                EOP_ButtonReleased
            };

            virtual void GetConfiguration( SFlowNodeConfig& config )
            {
                string joysticks = "enum_string:";
                int i = 1;

                for ( TJoysticks::const_iterator sit = g_joysticks.begin(); sit != g_joysticks.end(); ++sit )
                {
                    char buffer[32];
                    itoa( i, buffer, 10 );

                    joysticks.append( "Joystick" );
                    joysticks.append( buffer );
                    joysticks.append( "=" );
                    joysticks.append( buffer );
                    joysticks.append( "," );
                    i++;
                }

                static const SInputPortConfig inputs[] =
                {
                    InputPortConfig<bool>( "Enabled", true, _HELP( "Enables/Disables the joystick" ) ),
                    InputPortConfig<string>( "Joystick", "1", _HELP( "Select a joystick" ), "Joystick", _UICONFIG( joysticks ) ),
                    InputPortConfig_Null(),
                };

                static const SOutputPortConfig outputs[] =
                {
                    OutputPortConfig<int>( "X", _HELP( "X-axis, usually the left-right movement of a stick." ) ),
                    OutputPortConfig<int>( "Y", _HELP( "Y-axis, usually the forward-backward movement of a stick." ) ),
                    OutputPortConfig<int>( "Z", _HELP( "Z-axis, often the throttle control. If the joystick does not have this axis, the value is 0." ) ),
                    OutputPortConfig<int>( "X-Rotate", _HELP( "X-axis rotation. If the joystick does not have this axis, the value is 0." ) ),
                    OutputPortConfig<int>( "Y-Rotate", _HELP( "Y-axis rotation. If the joystick does not have this axis, the value is 0." ) ),
                    OutputPortConfig<int>( "Z-Rotate", _HELP( "Z-axis rotation (often called the rudder). If the joystick does not have this axis, the value is 0." ) ),
                    OutputPortConfig<int>( "Slider", _HELP( "Two additional axis values (formerly called the u-axis and v-axis) whose semantics depend on the joystick." ) ),
                    OutputPortConfig<int>( "POV", _HELP( "Direction controllers, such as point-of-view hats. The position is indicated in hundredths of a degree clockwise from north (away from the user). The center position is normally reported as - 1. For indicators that have only five positions, the value for a controller is - 1, 0, 9,000, 18,000, or 27,000." ) ),
                    OutputPortConfig<int>( "ButtonPressed", _HELP( "Joystick button pressed" ) ),
                    OutputPortConfig<int>( "ButtonReleased", _HELP( "Joystick button released" ) ),
                    OutputPortConfig_Null(),
                };

                config.pInputPorts = inputs;
                config.pOutputPorts = outputs;
                config.sDescription = _HELP( "Joystick infos" );
                config.SetCategory( EFLN_APPROVED );
            }

            void SetNewJoystick( int joystick )
            {
                if ( joystick == 0 )
                {
                    joystick = 1;
                }

                TJoysticks::iterator iter = g_joysticks.begin();

                if ( joystick == 1 && g_joysticks.size() == 1 )
                {
                    m_currentJoystick = iter->first;
                    m_CurrentCapabilities.dwSize = sizeof( DIDEVCAPS );
                    m_CurrentCapabilities = iter->second;

                    SAFE_DELETE( m_buttons );
                    m_buttons = new int[m_CurrentCapabilities.dwButtons];
                }

                else if ( g_joysticks.size() > 1 )
                {
                    TJoysticks::iterator end = g_joysticks.end();

                    int i = 2;
                    ++iter;

                    for ( ; iter != end; ++iter )
                    {
                        if ( i == joystick )
                        {
                            break;
                        }

                        i++;
                    }

                    m_currentJoystick = iter->first;
                    m_CurrentCapabilities.dwSize = sizeof( DIDEVCAPS );
                    m_CurrentCapabilities = iter->second;

                    SAFE_DELETE( m_buttons );
                    m_buttons = new int[m_CurrentCapabilities.dwButtons];
                }
            }

            virtual void ProcessEvent( EFlowEvent event, SActivationInfo* pActInfo )
            {
                switch ( event )
                {
                    case eFE_Initialize:
                        {
                            m_actInfo = *pActInfo;
                            m_bEnabled = true;
                            pActInfo->pGraph->SetRegularlyUpdated( pActInfo->myID, true );
                            SetNewJoystick( atoi( GetPortString( &m_actInfo, EIP_JoystickSelection ) ) );

                            break;
                        }

                    case eFE_Activate:
                        {
                            if ( IsPortActive( pActInfo, EIP_Enabled ) )
                            {
                                m_bEnabled = GetPortBool( pActInfo, EIP_Enabled );
                            }

                            else if ( IsPortActive( pActInfo, EIP_JoystickSelection ) )
                            {
                                SetNewJoystick( atoi( GetPortString( &m_actInfo, EIP_JoystickSelection ) ) );
                            }

                            break;
                        }

                    case eFE_Update:
                        {
                            if ( m_currentJoystick == NULL || m_buttons == NULL || !m_bEnabled )
                            {
                                return;
                            }

                            DIJOYSTATE2 js;

                            if ( PollJoystick( &js, m_currentJoystick ) == DI_OK )
                            {
                                m_xAxis = static_cast<int>( js.lX );
                                m_yAxis = static_cast<int>( js.lY );
                                m_zAxis = static_cast<int>( js.lZ );

                                m_xRotate = static_cast<int>( js.lRx );
                                m_yRotate = static_cast<int>( js.lRy );
                                m_zRotate = static_cast<int>( js.lRz );

                                m_slider = static_cast<int>( js.rglSlider[0] );
                                m_POV = static_cast<int>( js.rgdwPOV[0] );

                                ActivateOutput( pActInfo, EOP_JoystickX, m_xAxis );
                                ActivateOutput( pActInfo, EOP_JoystickY, m_yAxis );
                                ActivateOutput( pActInfo, EOP_JoystickZ, m_zAxis );

                                ActivateOutput( pActInfo, EOP_JoystickRotateX, m_xRotate );
                                ActivateOutput( pActInfo, EOP_JoystickRotateY, m_yRotate );
                                ActivateOutput( pActInfo, EOP_JoystickRotateZ, m_zRotate );

                                ActivateOutput( pActInfo, EOP_Slider, m_slider );
                                ActivateOutput( pActInfo, EOP_POV, m_POV );

                                for ( int i = 0; i < m_CurrentCapabilities.dwButtons; i++ )
                                {
                                    BYTE button = js.rgbButtons[i];

                                    if ( button & 0x80 )
                                    {
                                        if ( m_buttons[i] != 1 )
                                        {
                                            m_buttons[i] = 1;
                                        }

                                        ActivateOutput( pActInfo, EOP_ButtonPressed, i + 1 );
                                    }

                                    else
                                    {
                                        if ( m_buttons[i] == 1 )
                                        {
                                            m_buttons[i] = 0;
                                            ActivateOutput( pActInfo, EOP_ButtonReleased, i + 1 );
                                        }
                                    }
                                }
                            }

                            break;
                        }
                }
            }

            virtual void GetMemoryStatistics( ICrySizer* s )
            {
                s->Add( *this );
            }
    };
}

REGISTER_FLOW_NODE_EX( "Plugin_Joystick:Input", JoystickPlugin::CFlowJoystickNode, CFlowJoystickNode );